aws/signer/v4/middleware.go

```diff
@@ -55,13 +61,38 @@ func (m *UnsignedPayloadMiddleware) ID() string {
 func (m *UnsignedPayloadMiddleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (
 	out middleware.FinalizeOutput, metadata middleware.Metadata, err error,
 ) {
-	ctx = SetPayloadHash(ctx, v4Internal.UnsignedPayload)
+	req, ok := in.Request.(*smithyHTTP.Request)
+	if !ok {
+		return out, metadata, &HashComputationError{Err: fmt.Errorf("unexpected request middleware type %T", in.Request)}
+	}
+
+	contentSHA := v4Internal.UnsignedPayload
+
+	// Don't overwrite the content SHA if already set during serialization.
+	if vs := req.Header.Values(v4Internal.ContentSHAKey); len(vs) != 0 {
+		contentSHA = vs[0]
+	}
+
+	// Content SHA header only set when unsigned payload, unless otherwise
+	// customized, e.g. S3/Glacier
+	if contentSHA == v4Internal.UnsignedPayload {
+		req.Header.Set(v4Internal.ContentSHAKey, v4Internal.UnsignedPayload)
+	}
+
+	ctx = SetPayloadHash(ctx, contentSHA)
 	return next.HandleFinalize(ctx, in)
 }
 
 // ComputePayloadSHA256Middleware computes sha256 payload hash to sign
 type ComputePayloadSHA256Middleware struct{}
 
+// AddComputePayloadSHA256Middleware adds ComputePayloadSHA256Middleware to the
+// operation middleware stack
+func AddComputePayloadSHA256Middleware(stack *middleware.Stack) {
+	unsignedPayloadMiddleware := ComputePayloadSHA256Middleware{}
+	stack.Finalize.Add(&unsignedPayloadMiddleware, middleware.After)
```

Comment:
So it's Build or finalize? Computing Payload SHA can be build

Reply:
So this is one that I'd like to suggest both this and unsigned payload SHA256 middleware be moved form `Finalize` to `Build` to better signal that they are middleware that are only performed once. Where as middleware in `Finalize` could expect to be called multiple times per request attempt.

Reply:
Yeah I think both of them should be moved to Build.

Reply:
Also we should move setting the content-sha256 header to its own middleware probably